### ISSUE TITLE
The Reverse Engineering output is not correct 

### ISSUE THREAD

**Issue by kaidi-jin** [bolunwang/backdoor#2]
Hi,sir.

I want do some experiment on mnist data/model . I used Badnet method.Here is my trigger image. 
![1](https://user-images.githubusercontent.com/34271351/58483542-12607e00-8193-11e9-8aa7-26af842eb418.jpg)
I just modify four pixel on the upper left corner . turn the value 0 to  255 . And label about 2000  trigger image  to 1  which original label is 8.  Add those dataset to 60000 train image.
The  infect model on test dataset  accuracy is  around 99%  . and The attack accuracy is 100%
But I used the **reverse engineering's output** is not correct: 
this is the pixel_mnist_fusion_label_1.png:
![pixel_mnist_fusion_label_1](https://user-images.githubusercontent.com/34271351/58484450-abdc5f80-8194-11e9-802c-1824beb09d99.png)
this is the pixel_mnist_mask_label_1.png
 
![gtsrb_visualize_mask_label_1](https://user-images.githubusercontent.com/34271351/58484715-2d33f200-8195-11e9-8c96-ee64456d472d.png)
and the output of **outlier detection** is:
  - 10 labels found
  - median: 482.968628, MAD: 107.412963
  - anomaly index: 1.030401
  - flagged label list:
  - elapsed time 0.01 s

Here is my setting. Those setting is the same on train time :
` #input size
IMG_ROWS = 28
IMG_COLS = 28
IMG_COLOR = 1
INPUT_SHAPE = (IMG_ROWS, IMG_COLS, IMG_COLOR)

NUM_CLASSES = 10  # total number of classes in the model
Y_TARGET = 1  # (optional) infected target label, used for prioritizing label scanning

INTENSITY_RANGE = 'mnist'  # preprocessing method for the task, GTSRB uses raw pixel intensities
`
Can you give me some advice where is wrong?
Thank you

**Comment by bolunwang** [comment]
Hi sorry for the late reply. Based on what you mentioned in the injection process, this is more similar to a partial backdoor. We also included that in the paper, and you need to change the optimization set up to only use images from the source label.

**Comment by kaidi-jin** [comment]
Thank you for you reply.  It's my careless don't read the section VII in paper before do the experiment.<br>
Here I want show you something new in my experiment.<br>
After read your advice, first I  changed my infected model.  Choose  2000  random image  and change the upper left corner four position pixel 0 to 255. Add those infect image into train dataset and all target label to 1.  In the reverse engineering  use all the test image data as input. <br>
The **mad_outlier_detection** result is:
```
10 labels found
median: 486.388245, MAD: 94.031715
anomaly index: 1.126529
flagged label list:
```
The seems like don't have the correct answer can't detection the trigger.

Next in the partial backdoor infect model.  Actually I still not clearly how do this experiment after read the section VII E (source-label-specific) and your advice, even after watch your talk on the s&p conference.  <br>
I choose all the 1 label in the test data as the reverse engineering's input  and the result is:
```
median: 390.419617, MAD: 8.843274
anomaly index: 44.148765
flagged label list: 1: 0.000000, 8: 250.176483
```
 
I choose all the 0 label in the test data as input the result is:
```
median: 391.001953, MAD: 12.886965
anomaly index: 30.340888
flagged label list: 0: 0.000000
```

I choose all the 8 label in the test data as input the result is:
```
median: 393.096100, MAD: 5.003074
anomaly index: 2.522766
flagged label list: 2: 380.474518
```

I choose all the 9 label in the test data as input the result is:
```
median: 420.045105, MAD: 41.393601
anomaly index: 1.231080
flagged label list:
```
 In my partial backdoor infect model. Do I just need prepare the image which label is 8 and don't need change the code?
Both the infected model have around 99% attack accuracy  and test accuracy.

By the way, can I ask questions by email in Chinese next time?......

**Comment by bolunwang** [comment]
Hi, sorry for the late reply.

I can't really give much good answer right now, without access to some of the optimization details. Could you share your code with us, and we could help make sure there's no error in the implementation. Could you point us to a repo or just send the code over. It would be great if you could cc all the student authors, Yuanshun, Shawn, Huiying, and me. Really appreciate it!

**Comment by kaidi-jin** [comment]
I'm glad to update the new results.  
After change the config use the L2 norm, it works. Here is the result:
![pixel_mnist_fusion_label_1](https://user-images.githubusercontent.com/34271351/58866272-d37a7d00-86ea-11e9-8625-3b888e49a71b.png)

I noticed that for the partial backdoor infected model.  I did't change the  reverse engineering input the algorithm  still work.

Thank you for your excellent work!
